src/aspire/ctf/ctf_estimator.py

```diff
@@ -0,0 +1,877 @@
+            z = z + alpha2 * dz
+
+            if iter_no < 2:
+                stop_cond = np.minimum(np.minimum(dx, dy), dz) / 1000
+
+            if iter_no > 400:
+                stop_cond = np.maximum(np.maximum(dx, dy), dz) + 1
+
+            iter_no = iter_no + 1
+
+        df1 = (x + np.abs(y + z * 1j)) / 2
+        df2 = (x - np.abs(y + z * 1j)) / 2
+        angle_ast = np.angle(y + z * 1j) / 2  # Radians
+
+        inner_cosine = y * np.cos(2 * theta) + z * np.sin(2 * theta)
+        outer_sine = np.sin(a * x + a * inner_cosine - b)
+        outer_cosine = np.cos(a * x + a * inner_cosine - b)
+
+        sine_x_term = a
+        sine_y_term = a * np.cos(2 * theta)
+        sine_z_term = a * np.sin(2 * theta)
+
+        c1 = np.sum(np.abs(outer_sine) * signal)
+        c2 = np.sqrt(sum_A * np.sum(outer_sine ** 2))
+
+        p = c1 / c2
+
+        return df1, df2, angle_ast, p
+
+    # Note, This doesn't actually use anything from the class.
+    # It is used in a solver loop of some sort, so it may not be correct
+    # to just use what is avail in the obj.
+    def write_star(self, df1, df2, ang, cs, voltage, pixel_size, amp, name):
+        """
+        Writes starfile.
+        """
+
+        if os.path.isdir("results") is False:
+            os.mkdir("results")
+
+        f = open("results/" + os.path.splitext(name)[0] + ".log", "w")
+        f.write(
+            "data_root\n\nloop_\n_rlnMicrographName #1\n_rlnDefocusU #2\n_rlnDefocusV #3\n_rlnDefocusAngle #4\n"
+        )
+        f.write(
+            "_rlnSphericalAberration #5\n_rlnAmplitudeContrast #6\n_rlnVoltage #7\n_rlnDetectorPixelSize #8\n"
+        )
+        f.write(name)
+        f.write("\t")
+        f.write("%5.8f" % (df1))
+        f.write("\t")
+        f.write("%5.8f" % (df2))
+        f.write("\t")
+        f.write("%5.8f" % (ang))
+        f.write("\t")
+        f.write("%5.2f" % (cs))
+        f.write("\t")
+        f.write("%5.4f" % (amp))
+        f.write("\t")
+        f.write("%5.4f" % (voltage))
+        f.write("\t")
+        f.write("%5.4f" % (pixel_size))
+        f.close()
+
+
+def estimate_ctf(
+    data_folder,
+    pixel_size,
+    cs,
+    amplitude_contrast,
+    voltage,
+    num_tapers,
+    psd_size,
+    g_min,
+    g_max,
+    output_dir,
+    dtype=np.float32,
+):
+    """
+    Given paramaters estimates CTF from experimental data
+    and returns CTF as a mrc file.
+    """
+
+    dtype = np.dtype(dtype)
+    assert dtype in (np.float32, np.float64)
+
+    dir_content = os.scandir(data_folder)
+
+    mrc_files = [f.name for f in dir_content if os.path.splitext(f)[1] == ".mrc"]
+    mrcs_files = [f.name for f in dir_content if os.path.splitext(f)[1] == ".mrcs"]
+    file_names = mrc_files + mrcs_files
+
+    amp = amplitude_contrast
+    amplitude_contrast = np.arctan(
+        amplitude_contrast / np.sqrt(1 - np.square(amplitude_contrast))
+    )
+    lmbd = 1.22639 / np.sqrt(voltage * 1000 + 0.97845 * np.square(voltage))
```

Comment:
`voltage_to_wavelength`

Reply:
changed this one (voltage_to_wavelength / 10).  Also found last time this was replaced (probably me) it wasn't done right, so I fixed that too.

```
self.lmbd = voltage_to_wavelength(voltage) / 10
print(1.22639 / np.sqrt(voltage * 1000 + 0.97845 * np.square(voltage)), self.lmbd)
```

```
0.0019687615309233577 0.0019686970075614534
```

I'm guessing this is a wavelength nm/angstrom unit diff?.. so I went ahead and labeled the units in the code a little more.

Reply:
Looks good. Yes it would make sense that it's converting from angstrom to nm.